### PR TITLE
[PS-18846] Allow Bors to use local merge

### PR DESCRIPTION
**Pull request by wyao-simspace**
[PS-18846](https://ticket.simspace.com/browse/PS-18846)

Adds a configuration flag that makes Bors use CLI Git locally to merge batches of PRs together, and implements the functionality to do said local merge. Additionally, adds the ability for Bors to run shell script hooks before and after the batch merge process, and including any changes made by said hooks into the staging branch.

## An overview of how the code actually works

For our purposes the toplevel module is `batcher.ex`. It's responsible for the entire lifetime of a batch of PRs.

Each batch of PRs can be in a few different states, including Running and Waiting to run. When you comment `bors r+` on a PR, a webhook eventually makes its way into `batcher.ex`, either adding the PR into the current waiting batch or creating a new one if there is no waiting batch. In the background `batcher.ex` runs a polling loop to pick up waiting batches after finishing up its previous batch; upon finding one, it calls `start_waiting_batch()` to merge the batched PRs with your master branch and start CI.

Specifically, `start_waiting_batch()` did the following:

+ Merge master + all the PR branches into `staging.tmp` one-by-one.
+ *Manually* create a commit using the *contents* of `staging.tmp`, ignoring that branch's commit history. That way it looks like Bors created only a single merge commit.
+ Force-push `staging` to that newly-created commit

Crucially, once `staging.tmp` has the right contents, there's nothing further that needs to be done locally; we only need a function that handles up through merging the PRs locally and pushing the local repo's contents up to `staging.tmp`. Structurally, I split out the current merging/pushing behavior into a separate module `BorsNG.GitHub.Merge.API`, and a new module `BorsNG.GitHub.Merge.Local` that provides the same functions, but using CLI Git. `start_waiting_batch()` then decides which module's functions to call into using the configuration flag.

Once the correct contents are on the `staging` branch, Bors doesn't care about the repo contents ever again. So from there, we can simply rely on all the existing Bors machinery; CI success on `staging` will cause Bors to fast-forward the master branch to the current state of `staging`. CI failure will cause Bors to bisect the failed batch, creating two new batches which will both eventually get passed to `start_waiting_batch()`.

### Squash merging

Bors has a flag in the `bors.toml` project config called `use_squash_merge`, which squashes each PR down to a single commit when merging batches together. If squash merging is enabled, Bors does this squash merging *after* doing the normal merge described above. This might seem weird, but it's intentional so Bors can pull a canonical `bors.toml` from the merge result before deciding to squash merge; Bors wants to allow PRs to make changes to the project configuration.

I added a similar function for doing the squash merge, and inside `start_waiting_merged_batch()` (which is called by `start_waiting_batch()`), it does the same conditional selection of which module function to call.

### CODE REVIEW

File: lib/github/merge/local.ex [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  alias BorsNG.Database.Batch
+  alias BorsNG.Database.Project
+  alias BorsNG.Database.LinkPatchBatch
+  alias BorsNG.GitHub
+  alias BorsNG.GitHub.Merge.Hooks
+  alias BorsNG.Worker.Batcher
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
+  end
```
Comment (wyao-simspace) [inline lib/github/merge/local.ex:21]: about GenServer: this is an Erlang-y/Elixir-y idiom. rather than this being a module that you call library functions from, this module is actually a separate (lightweight Erlang) process that receives messages from other processes. if it crashes or exceptions happen, this process will automatically get restarted without affecting the rest of the application. it doesn't change the code that much, but the main thing this enables is that exceptions or pattern match failures aren't as important, since it won't cause loss of service.

importantly, the code is written in such a way that there's only one process based off this module running at any given time, since we don't want multiple things modifying the filesystem at the same time.

File: lib/github/merge/local.ex [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+            git.(["fetch", "origin", patch.commit])
+
+            {_, exit_code} =
+              git.([
+                "merge",
+                patch.commit,
+                "-m",
+                "[ci skip][skip ci][skip netlify] -bors-staging-tmp-#{patch.pr_xref}"
+              ])
```
Comment (wyao-simspace) [inline lib/github/merge/local.ex:93]: note that the commit message here is completely irrelevant: Bors only cares about the *contents* of the `staging.tmp` branch, which it uses to construct the actual merge commit manually. so these commit messages will never get seen. we at least set ci skip to avoid cluttering CI with irrelevant runs on the `staging.tmp` branch.

File: lib/github/merge/local.ex [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        # Manually create a squash commit with the directory contents after the merge
+        commit_message =
+          Batcher.Message.generate_squash_commit_message(
+            pr,
+            commits,
+            user_email,
+            toml.cut_body_after
+          )
+
+        {new_head, 0} = git.(["commit-tree", tree, "-p", prev_head, "-m", commit_message])
```
Comment (wyao-simspace) [inline lib/github/merge/local.ex:169]: here, when doing a local squash merge, we actually *do* have to care about the commit message, as each PR will become a separate commit, and Bors will avoid adding an additional merge commit on top; these commits do eventually get seen on master. the construction here is a little subtle; we're manually constructing a commit through the CLI by taking the repo contents from `git merge` (in `tree`), but with master as the previous commit in the history (in `prev_head`).

File: lib/worker/batcher.ex [hunk 4]
```diff
@@ -466,92 +436,25 @@ defmodule BorsNG.Worker.Batcher do
...
+              commit = Merge.API.squash_merge_batch!(batch, patch_links, base, toml)
+              [commit]
+
+            not squash? ->
+              [base.commit | Enum.map(patch_links, & &1.patch.commit)]
           end
 
         head =
-          if toml.use_squash_merge do
+          if squash? do
```
Comment (wyao-simspace) [inline lib/worker/batcher.ex:457]: below here is where Bors is actually pushing staging to the merge result. again, note that Bors manually constructs the merge commit with `synthesize_commit!()` so that there's only a single merge commit per batch (unless squash merges are enabled)

**Comment by wyao-simspace** [comment]
I wonder if it would be best to allow local merge to be specified on a per-project basis in `bors.toml`. The thing is that we'd still have to have the global config flag as well, so that if we eventually upstream these changes, upstream Bors/other deployments of Bors can ignore that part of the TOML. It's a bit more work for no real benefit on our end right now.

File: config/config.exs [hunk 5]
```diff
@@ -56,6 +56,9 @@ config :bors, BorsNG.Endpoint,
   render_errors: [view: BorsNG.ErrorView, accepts: ~w(html json)],
   pubsub: [name: BorsNG.PubSub, adapter: Phoenix.PubSub.PG2]
 
+# Configure hooks when local merge is enabled
+config :bors, BorsNG.GitHub.Merge.Hooks, hooks_dir: {:system, "HOOKS_DIR", ".bors-hooks"}
```
Comment (dfithian) [inline config/config.exs:60]: The comment doesn't make sense to me here. There's nothing that describes a local merge in this line?

File: config/prod.exs [hunk 6]
```diff
@@ -3,4 +3,6 @@ use Mix.Config
 config :bors, :server, BorsNG.GitHub.Server
 config :bors, :oauth2, BorsNG.GitHub.OAuth2
 
+config :bors, :local_merge?, {:system, :boolean, "USE_LOCAL_MERGE", false}
```
Comment (dfithian) [inline config/prod.exs:6]: Which one of these environments do we use for the portal-suite repo?

File: config/config.exs (hunk 5 above)
Comment (wyao-simspace) [reply config/config.exs:60]: shell script hooks are only used when local merge is enabled; this lets a deployment configure where those hooks are stored in a repo.

this is another config value that might be better as a per-project config value in `bors.toml` though

File: config/prod.exs (hunk 6 above)
Comment (wyao-simspace) [reply config/prod.exs:6]: prod; dev and test don't hit GitHub at all

File: lib/github/merge/api.ex [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    tbase = %{
+      tree: base.tree,
+      commit:
+        GitHub.synthesize_commit!(
+          repo_conn,
+          %{
+            branch: stmp,
+            tree: base.tree,
+            parents: [base.commit],
+            commit_message: "[ci skip][skip ci][skip netlify]",
```
Comment (dfithian) [inline lib/github/merge/api.ex:34]: This is intending to skip CI when the commit is merged into the branch? Did you test this to make sure it works?

Comment (wyao-simspace) [reply lib/github/merge/api.ex:34]: yes, i tested it manually. i'm still writing the more automated tests

this module is copied straight from master bors; you can see this code already existed in `batcher.ex`, i just moved it out. but for why Bors is doing this in the first place, note that this function (`merge_batch!`) merges into *staging.tmp*, not staging proper, so CI is irrelevant on this branch. all the commit construction happening here is a bit subtle, because the commits created in `merge_batch!()` are never actually visible! at the bottom of `start_waiting_merged_batch()` in `batcher.ex` is where staging proper actually gets updated; you'll see that Bors completely ignores any commits created by this function and constructs a new one with the correct commit message, parented onto the batch's `into_branch` (dev in our case)

File: lib/github/merge/hooks.ex [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  defp hooks_dir do
+    config()[:hooks_dir]
+  end
+
+  @spec invoke_before_merge_hook!(binary) :: nil
+  def invoke_before_merge_hook!(project_dir) do
+    hook_file =
+      File.cwd!()
+      |> Path.join(project_dir)
+      |> Path.join(hooks_dir())
```
Comment (alec-simspace) [inline lib/github/merge/hooks.ex:21]: What happens if `config()[:hooks_dir]` isn't set?

Comment (wyao-simspace) [reply lib/github/merge/hooks.ex:21]: it defaults to `.bors-hooks/`. you can see the default value in `config.exs`.

**Comment by alec-simspace** [comment]
rockin', thanks!

File: lib/github/merge/local.ex [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    git.(["fetch", "origin", batch.into_branch])
+    git.(["checkout", "origin/#{batch.into_branch}"])
+
+    Hooks.invoke_before_merge_hook!(workdir)
+
+    merge_status =
+      patch_links
+      |> Enum.reduce(:merged, fn link_patch_batch, status ->
+        case status do
+          :conflict ->
```
Comment (Sciencei) [inline lib/github/merge/local.ex:94]: do we need to handle any other statuses besides conflict and not-conflict? the api one also handles canceled

File: lib/github/merge/local.ex [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+    # the .git/config.
+    File.rm_rf!(workdir)
+    init_project_repo!(batch.project, cli_creds(repo_conn))
+
+    git = fn args -> System.cmd("git", @bors_author ++ args, cd: workdir) end
+
+    git.(["fetch", "origin", batch.into_branch])
+    git.(["checkout", "origin/#{batch.into_branch}"])
+
+    Hooks.invoke_before_merge_hook!(workdir)
```
Comment (Sciencei) [inline lib/github/merge/local.ex:151]: nit: duplicated code, may be worth factoring out

File: lib/github/merge/local.ex [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+    :ok
+  end
+
+  @spec persist_local_to_github!(String.t(), String.t()) :: :ok
+  defp persist_local_to_github!(branch_name, workdir) do
+    # Credentials already exist in the URL for 'origin' here, from the clone.
+    # Each access token is valid for an hour, so there shouldn't be a need to
+    # refresh it before pushing.
+    {_, 0} =
```
Comment (Sciencei) [inline lib/github/merge/local.ex:225]: while generally unlikely, spuriously failing the batch is definitely bad. I think it's probably OK enough for our local fork, but should be fixed before upstreaming

File: lib/github/merge/local.ex [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+  @doc """
+  The intended semantics are that hooks can change the repo filesystem, and
+  we'll combine those changes with our merge commit. This function is careful
+  to not add a new commit to the commit history.
+  """
+  @spec commit_hook_changes!(String.t()) :: :ok
+  def commit_hook_changes!(workdir) do
+    git = fn args -> System.cmd("git", @bors_author ++ args, cd: workdir) end
+    {last_msg, 0} = git.(["log", "-n1", "--pretty=%B"])
+    git.(["add", "-A"])
```
Comment (Sciencei) [inline lib/github/merge/local.ex:244]: check that this captures deletions

File: lib/web/webhook_parser_plug.ex [hunk 13]
```diff
@@ -23,7 +23,7 @@ defmodule BorsNG.WebhookParserPlug do
   end
 
   def run(conn, _options, key) do
-    {:ok, body, _} = read_body(conn)
+    {:ok, body, conn} = read_body(conn)
```
Comment (Sciencei) [inline lib/web/webhook_parser_plug.ex:26]: I would keep this in a separate commit if it isn't already

File: lib/github/merge/hooks.ex [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  defp config do
+    Confex.fetch_env!(:bors, __MODULE__)
+  end
+
+  defp hooks_dir do
+    config()[:hooks_dir]
+  end
+
+  @spec invoke_before_merge_hook!(binary) :: nil
+  def invoke_before_merge_hook!(project_dir) do
```
Comment (Sciencei) [inline lib/github/merge/hooks.ex:17]: nit: these functions are almost identical, could be worth factoring them out

**Review by Sciencei — CHANGES_REQUESTED** [review]
a couple small things that I'd like looked at, largely LGTM though

File: lib/github/merge/local.ex (hunk 12 above)
Comment (wyao-simspace) [reply lib/github/merge/local.ex:244]: tested, captures changes, new files, and file deletions

File: lib/github/merge/local.ex (hunk 9 above)
Comment (wyao-simspace) [reply lib/github/merge/local.ex:94]: weirdly enough, I can't find any way the equivalent API merge code could receive a :canceled. none of the functions it calls return that. so that case seems to be a dead branch in the API merge?

for example, here's the signature of `merge_branch!`

```elixir
  @spec merge_branch!(tconn, %{
          from: bitstring,
          to: bitstring,
          commit_message: bitstring
        }) :: %{commit: binary, tree: binary} | :conflict
```

very strange.

**Review by Sciencei — APPROVED** [review]
